### DISCUSSION TITLE
Can't re-use pv (retain) after deleting vela app

### DISCUSSION THREAD

**Discussion by hprotzek** [kubevela/kubevela#1574] [Q&A]
Hi there, I am trying to create a gcs bucket trait with the `csi-gcs` driver. Using existing buckets, forces me to use `Retain` policy on the pv. When I delete the vela app, the pc get in state `Released` but still have the old owner ref, which prevents using it with a new `vela up ...`

```
cannot apply trait "v1" "PersistentVolume" "csi-gcs-pv-snbucket-test": cannot apply ApplyOption: existing object is not controlled by UID "c4875aba-b301-4072-a388-5cc4e4013be8
```

```
# Code generated by KubeVela templates. DO NOT EDIT.
apiVersion: core.oam.dev/v1beta1
kind: TraitDefinition
metadata:
  name: snbucket
  annotations:
    definition.oam.dev/description: Mount a Google Cloud Storage bucket
spec:
  appliesToWorkloads:
    - snweb
  schematic:
    cue:
      template: |
        parameter: {
        	bucketName: string
        	// +usage=Mount path in container
        	mountPath: string
        	// +usage=Service account json
        	serviceAccount: string
        }
        patch: {
        	spec: template: spec: {
        		// +patchKey=name
        		containers: [{
        			name: context.name
        			// +patchKey=name
        			volumeMounts: [{
        				name:      "bucket1"
        				mountPath: parameter.mountPath
        			}]
        		}]
        		// +patchKey=name
        		volumes: [{
        			name: "bucket1"
        			persistentVolumeClaim: claimName: "csi-gcs-pvc-\(context.name)"
        		}]
        	}
        }
        outputs: {
        	secret: {
        		apiVersion: "v1"
        		kind:       "Secret"
        		metadata: name: "bucket-sa-\(context.name)"
        		type: "Opaque"
        		stringData: {
        			bucket: parameter.bucketName
        			key:    parameter.serviceAccount
        		}
        	}
        
        	pv: {
        		apiVersion: "v1"
        		kind:       "PersistentVolume"
        		metadata: name: "csi-gcs-pv-\(context.name)"
        		spec: {
        			accessModes: ["ReadWriteOnce"]
        			capacity: storage: "5Gi"
        			persistentVolumeReclaimPolicy: "Retain"
        			storageClassName:              "csi-gcs-test-sc"
        			csi: {
        				driver:       "gcs.csi.ofek.dev"
        				volumeHandle: "csi-gcs"
        				nodePublishSecretRef: {
        					name:      "bucket-sa-\(context.name)"
        					namespace: context.namespace
        				}
        			}
        		}
        	}
        
        	pvc: {
        		apiVersion: "v1"
        		kind:       "PersistentVolumeClaim"
        		metadata: name: "csi-gcs-pvc-\(context.name)"
        		spec: {
        			accessModes: ["ReadWriteOnce"]
        			resources: requests: storage: "5Gi"
        			storageClassName: "csi-gcs-test-sc"
        		}
        	}
        }
 ```

Any idea how to solve that? I could generate a random name for the pv, but CUE doesn't provide any random functions, what I see so far? Or could vela remove the owner ref?

**Comment by captainroy-hy** [comment]
After investigation, I cannot figure out a solution or workaround for this issue. AppContext(AppConfig) controller always assumes that a resource being applied must be controlled by the Application and will be deleted cascadingly with the deletion of Application.
In this case, a PersistentVolume with a `Retain` policy will not be deleted even its controller Application is deleted.
@wonderflow A possible solution is, during GC in AppConfig controller, we remove the controller owner before deleting resources.

**Reply by wonderflow** [reply]
Discussed offline,  @captainroy-hy will create an issue to fix it. KubeVela will check the owner's existence before apply.

**Comment by wonderflow** [comment]
Let's convert it to be an issue and discuss more  https://github.com/oam-dev/kubevela/issues/1614

**Reply by hprotzek** [reply]
@captainroy-hy @wonderflow thank you for picking this up, really appreciated :-) also happy to help to implement or test it

**Reply by wonderflow** [reply]
@hprotzek Hi friends, I think we have already fixed this issue #1634  please let me know if you have any feedback. BTW,  could you help to finish the survey if you're using KubeVela https://github.com/oam-dev/kubevela/issues/1662

**Reply by hprotzek** [reply]
Yes it works now, thank you very much for your work. Will add us to the survey too :-)
